### ISSUE TITLE
Google OAuth client can’t handle application_default_credentials.json

### ISSUE THREAD

**Issue by yath** [collectd/collectd#3540]
The documentation for `write_stackdriver` [says](https://github.com/collectd/collectd/blob/main/src/collectd.conf.pod#plugin-write_stackdriver):

> The path `${HOME}/.config/gcloud/application_default_credentials.json` is checked. This where credentials used by the gcloud command line utility are stored. You can use `gcloud auth application-default login` to create these credentials.

Ok, so I do:

```
% gcloud auth application-default login                              
Your browser has been opened to visit:

    https://accounts.google.com/o/oauth2/auth?client_id=[…]&access_type=offline&response_type=code&prompt=select_account

Credentials saved to file: [/home/yath/.config/gcloud/application_default_credentials.json]

These credentials will be used by any library that requests Application Default Credentials (ADC).
```

And add to my `collectd.conf`:
```
<Plugin write_stackdriver>
   Project "my-gcp-project"
</Plugin>
```

However startup fails with `utils_oauth: oauth_create_google_json: project_id field not found`.

My `application_default_credentials.json` indeed has no such field, neither does it have `client_email`, `token_uri` or `private_key`, as attempted to parse further down in [this function](https://github.com/collectd/collectd/blob/main/src/utils/oauth/oauth.c#L471).

My `application_default_credentials.json` looks more like this:

```json
{
  "client_id": "1234567890-....apps.googleusercontent.com",
  "client_secret": "d-...",
  "refresh_token": "1//...",
  "type": "authorized_user"
}
```

Where the refresh token can be [used](https://developers.google.com/identity/protocols/oauth2) to request a short-lived access token. The code looks like it’s attempting to parse a service account key file, which probably is different but I can’t test because my GCP organization forbids creation of such key pairs.

**Comment by octo** [comment]
Right, looks like we need to treat `application_default_credentials.json` differently from the service account keys.

IIuc, we need to use the [Refreshing an access token](https://developers.google.com/identity/protocols/oauth2/native-app#offline) call. The response can be parsed by `oauth_parse_json_token()`.

To implement, I suggest the following steps:

1.  Extend `struct oauth_s` aka. `oauth_t` to also hold:
    *   `client_id`
    *   `client_secret`
    *   `refresh_token`
    *   A "type" field that can be used to trigger either the JWT or refresh token implementation.
1.  Implement a new function, `new_refresh_token()` similar to `new_token()` which is using the [Refreshing an access token](https://developers.google.com/identity/protocols/oauth2/native-app#offline) flow.
1.  Rename `new_token()` to `new_jwt_token()`.
1.  Extend `renew_token()` to call either `new_jwt_token()` or `new_refresh_token()` depending on the new "type" field.
1.  Add refresh-token specific variants of `oauth_create_google_json()` and `oauth_create_google_file()`.
1. Update `oauth_create_google_default()` to parse the file with the new refresh-token file parsing function.

Does that sound about right?

**Comment by yath** [comment]
> Implement a new function, new_refresh_token() similar to new_token() which is using the Refreshing an access token flow.

Terminology nit: The refresh token stays the same (it’s the long-term secret generated by the OAuth2 consent flow), the access token expires. So I’d call it `new_access_token()`; it just utilizes the refresh token for refreshing the access token.

Otherwise LGTM, thanks!

**Comment by octo** [comment]
Discussed in-person: I suggested "refresh" to differentiate from the "jwt" specific code. "new_refresh_token" is very misleading and not a good name though.

Since [RFC 8252](https://tools.ietf.org/html/rfc8252) talks about "native applications", I suggest to use "native" and "jwt" to make the distinction, e.g. `new_token_jwt` and `new_token_native`.
